src/Lib.P2P/PubSub/NotificationService.cs

```diff
@@ -0,0 +1,242 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Logging;
+
+namespace Lib.P2P.PubSub
+{
+    /// <summary>
+    ///   A simple pub/sub messaging service that supports
+    ///   multiple message routers.
+    /// </summary>
+    /// <remarks>
+    ///   Relies upon the router(s) to deliver and receive messages from other peers.
+    /// </remarks>
+    public class NotificationService : IService, IPubSub
+    {
+        private static ILog log = LogManager.GetLogger(typeof(NotificationService));
+
+        private class TopicHandler
+        {
+            public string Topic;
+            public Action<IPublishedMessage> Handler;
+        }
+
+        private long nextSequenceNumber;
+        private ConcurrentDictionary<TopicHandler, TopicHandler> topicHandlers;
+        private MessageTracker tracker = new MessageTracker();
+
+        // TODO: A general purpose CancellationTokenSource that stops publishing of
+        // messages when this service is stopped.
+
+        /// <summary>
+        ///   The local peer.
+        /// </summary>
+        public Peer LocalPeer { get; set; }
+
+        /// <summary>
+        ///   Sends and receives messages to other peers.
+        /// </summary>
+        public List<IMessageRouter> Routers { get; set; } = new List<IMessageRouter>
+        {
+            new LoopbackRouter()
+        };
+
+        /// <summary>
+        ///   The number of messages that have published.
+        /// </summary>
+        public ulong MesssagesPublished;
+
+        /// <summary>
+        ///   The number of messages that have been received.
+        /// </summary>
+        public ulong MesssagesReceived;
+
+        /// <summary>
+        ///   The number of duplicate messages that have been received.
+        /// </summary>
+        public ulong DuplicateMesssagesReceived;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)